receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processscraper
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/process"
+	"go.opencensus.io/trace"
+
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/internal/processor/filterset"
+)
+
+// scraper for Process Metrics
+type scraper struct {
+	config    *Config
+	startTime pdata.TimestampUnixNano
+	includeFS filterset.FilterSet
+	excludeFS filterset.FilterSet
+}
+
+type processMetadata struct {
+	pid      int32
+	name     string
+	handle   processHandle
+	username string
+	cmdline  string
+}
+
+type processTimes struct {
+	*processMetadata
+	times *cpu.TimesStat
+}
+
+type processMemoryInfo struct {
+	*processMetadata
+	memoryInfo *process.MemoryInfoStat
+}
+
+type processIoCounters struct {
+	*processMetadata
+	ioCounters *process.IOCountersStat
+}
+
+// newProcessScraper creates a Process Scraper
+func newProcessScraper(_ context.Context, cfg *Config) (*scraper, error) {
+	scraper := &scraper{config: cfg}
+
+	var err error
+
+	if len(cfg.Include.Names) > 0 {
+		scraper.includeFS, err = filterset.CreateFilterSet(cfg.Include.Names, &cfg.Include.Config)
+		if err != nil {
+			return nil, fmt.Errorf("error creating process include filters: %v", err)
+		}
+	}
+
+	if len(cfg.Exclude.Names) > 0 {
+		scraper.excludeFS, err = filterset.CreateFilterSet(cfg.Exclude.Names, &cfg.Exclude.Config)
+		if err != nil {
+			return nil, fmt.Errorf("error creating process exclude filters: %v", err)
+		}
+	}
+
+	return scraper, nil
+}
+
+// Initialize
+func (s *scraper) Initialize(_ context.Context) error {
+	bootTime, err := host.BootTime()
+	if err != nil {
+		return err
+	}
+
+	s.startTime = pdata.TimestampUnixNano(bootTime)
+	return nil
+}
+
+// Close
+func (s *scraper) Close(_ context.Context) error {
+	return nil
+}
+
+// ScrapeMetrics
+func (s *scraper) ScrapeMetrics(ctx context.Context) (pdata.MetricSlice, error) {
+	_, span := trace.StartSpan(ctx, "processscraper.ScrapeMetrics")
+	defer span.End()
+
+	var errors []error
+	metrics := pdata.NewMetricSlice()
+
+	processes, err := s.getProcesses()
+	if len(processes) == 0 {
+		return metrics, err
+	}
+
+	if err != nil {
```

Comment:
This condition should be impossible given the check for empty processes above, since result must be empty when err is not nil.
Is there a possibility of getting non-empty list of processes AND error at the same time?

Reply:
See the comment / discussion above which explains the reasoning for this. Note the additional check that was here before for processes being empty wasn't necessary. I've changed this now to simply read:

```go
processes, err := s.getProcesses()
if err != nil {
	errors = append(errors, err)
}
```

... which is closer to being idiomatic

Reply:
```go
if err := scrapeAndAppendCPUUsageMetric(metrics, s.startTime, processes); err != nil {
  errors = append(errors, err)
}
```

Same everywhere where you have just an error. It is more compact and simpler to read (personal opinion).

Reply:
Updated